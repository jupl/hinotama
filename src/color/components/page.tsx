import {inject, observer} from 'mobx-react'
import * as React from 'react'
import styled from 'styled-components'
import {IColor} from '../stores'

// Background component
interface BackgroundProps {
  color: string
  className?: string
  children?: React.ReactNode
}
const Background = ({color: _, ...props}: BackgroundProps) =>
  <div {...props} />
const StyledBackground = styled(Background)`
  transition: background 0.8s ease-out;
  display: flex;
  background-color: ${({color}) => color};
`

// Text component
const Text = styled.span`
  margin-left: 4px;
  margin-right: 4px;
`

// Gradient overlay component
const Gradient = styled.div`
  flex: 1;
  display: flex;
  justify-content: center;
  align-items: center;
  background: linear-gradient(rgba(255, 255, 255, 0.4), transparent);
`

/** Button component */
export const Button = styled.button`
  border-radius: 50%;
  background: radial-gradient(white, gainsboro);
  border-color: gainsboro;
  outline: none;
`

/** Properties for color page component */
export interface Props {
  /** CSS color to show */
  readonly color?: IColor

  /** Additional class names to include */
  readonly className?: string
}

// Bind color page component with properties from Mobx provider
export default inject('color')(observer(ColorPage))

/**
 * Render color page component showing current color and buttons to change
 * @param props Component properties
 * @return Color page component
 */
export function ColorPage({className, color}: Props) {
  // Force non-undefined stores due to behavior of Mobx's inject
  color = color!

  return (
    <StyledBackground color={color.value} className={className}>
      <Gradient>
        <Button onClick={color.previousColor}>&lt;</Button>
        <Text>Hello, World</Text>
        <Button onClick={color.nextColor}>&gt;</Button>
      </Gradient>
    </StyledBackground>
  )
}
