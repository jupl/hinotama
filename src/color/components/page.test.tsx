import * as React from 'react'
import {shallow} from 'enzyme'
import {Button, ColorPage} from './page'

// tslint:disable:no-magic-numbers

describe('<Color.Page> Template', () => {
  const mockColor = {
    value: 'white',
    previousColor: jest.fn(),
    nextColor: jest.fn(),
  }

  it('should render as expected', () => {
    expect(shallow(<ColorPage color={mockColor} />)).toMatchSnapshot()
  })

  it('should invoke events as expected', () => {
    const component = shallow(<ColorPage color={mockColor} />)
    const buttons = component.find(Button)
    const previousButton = buttons.at(0)
    const nextButton = buttons.at(1)

    expect(buttons.length).toBe(2)
    expect(mockColor.previousColor).not.toBeCalled()
    expect(mockColor.nextColor).not.toBeCalled()

    previousButton.simulate('click')
    expect(mockColor.previousColor.mock.calls.length).toBe(1)
    expect(mockColor.nextColor).not.toHaveBeenCalled()

    nextButton.simulate('click')
    expect(mockColor.previousColor.mock.calls.length).toBe(1)
    expect(mockColor.nextColor.mock.calls.length).toBe(1)
  })
})
