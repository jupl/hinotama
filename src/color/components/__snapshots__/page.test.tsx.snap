// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Color.Page> Template should render as expected 1`] = `
<Styled(Background)
  color="white"
>
  <styled.div>
    <styled.button
      onClick={[Function]}
    >
      &lt;
    </styled.button>
    <styled.span>
      Hello, World
    </styled.span>
    <styled.button
      onClick={[Function]}
    >
      &gt;
    </styled.button>
  </styled.div>
</Styled(Background)>
`;
